File: reverseLinkedList.js
const reverseList = function(head){
	if (!head || !head.next) {
		return head;
	}
	let previous = null;
	let current = head;
	let next = current.next;
	while(current){
		current.next = previous;
		previous = current;
		current = next;
		next = next ? next.next : null;
	}
	return previous;
}

const ListNode = function(val) {
	this.val = val;
	this.next = null;
}

const test = new ListNode(1);
const reversed = new ListNode(4);
test.next = new ListNode(2);
test.next.next = new ListNode(3);
test.next.next.next = new ListNode(4);
reversed.next = new ListNode(3);
reversed.next.next = new ListNode(2);
reversed.next.next.next = new ListNode(1);
const single = new ListNode(6);

console.assert(JSON.stringify(reverseList(test)) === JSON.stringify(reversed), "reverses a linked list");
console.assert(JSON.stringify(reverseList(single)) === JSON.stringify(single), "reverse a single node");
console.assert(JSON.stringify(reverseList(null)) === JSON.stringify(null), "works with a null head");

File: reverseLinkedList.py
# https://leetcode.com/problems/reverse-linked-list/

class ListNode:
	def __init__(self, val = 0, next = None):
		self.val = val
		self.next = next

class Solution:
	def reverseList(self, head: ListNode) -> ListNode:
		prev = None
		next = head
		while next != None:
			temp = next.next
			next.next = prev
			prev = next
			next = temp
		return prev

test = Solution()
case1 = ListNode(1, ListNode(2, ListNode(3, ListNode(4))))
res1 = test.reverseList(case1)

print(res1.val, res1.next.val, res1.next.next.val, res1.next.next.next.val)
